File: contactApp.cpp
/* contacts web microservice using mariaDB
 * By James Skon, Kenyon College, 2022
 */


#include <iostream>
#include <fstream>
#include <string>
#include <map>

#include "httplib.h"
#include "contactDB.h"
#include "contactEntry.h"
const int port = 5004;

using namespace std;

ofstream logfile; 

string jsonResults(vector<contactEntry> pbList) {
	string res = "{\"results\":[";
	for (int i = 0; i<pbList.size(); i++) {
		res += pbList[i].json();
		if (i < pbList.size()-1) {
			res +=",";
		}
	}
	res += "]}";
	return res;
}

int main() {
	httplib::Server svr;

  	contactDB ctdb; // Contact Book SQL Interface Object
  
  	vector<contactEntry> results;

  	svr.Get("/", [](const httplib::Request & /*req*/, httplib::Response &res) {
    	res.set_header("Access-Control-Allow-Origin","*");
    	res.set_content("Contact API", "text/plain");
  	});

  	svr.Get(R"(/contact/find)", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	results = ctdb.find("");
    	string json = jsonResults(results);
    	res.set_content(json, "text/json");
    	res.status = 200;
  	});
  	  	
  	svr.Get(R"(/contact/find/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string last = req.matches[1];
    	results = ctdb.find(last);
    	string json = jsonResults(results);
    	res.set_content(json, "text/json");
    	res.status = 200;
  	});
  	
  	svr.Get(R"(/contact/last/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string last = req.matches[1];
    	results = ctdb.findByLast(last);
    	string json = jsonResults(results);
    	cout << "Last: " << json << endl;
    	res.set_content(json, "text/json");
    	res.status = 200;
  	});
  	
  	svr.Get(R"(/contact/first/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string first = req.matches[1];
    	results = ctdb.findByFirst(first);
    	string json = jsonResults(results);
    	cout << "first: " << json << endl;
    	res.set_content(json, "text/json");
    	res.status = 200;
  	});  	

  	svr.Get(R"(/contact/type/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string type = req.matches[1];
    	results = ctdb.findByType(type);
    	string json = jsonResults(results);
    	cout << "type: " << json << endl;
    	res.set_content(json, "text/json");
    	res.status = 200;
  	});  
  	
  	svr.Get(R"(/contact/add/(.*)/(.*)/(.*)/(.*)/(.*)/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string first = req.matches[1];
    	string last = req.matches[2];
    	string phone = req.matches[3];
    	string type = req.matches[4];
    	string email = req.matches[5];
    	string age = req.matches[6];
    	ctdb.addEntry(first,last,phone,type,email,age);

    	res.set_content("{\"status\":\"success\"}", "text/json");
    	res.status = 200;
  	}); 	
 
   	svr.Get(R"(/contact/update/(.*)/(.*)/(.*)/(.*)/(.*)/(.*)/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string ID = req.matches[1];
    	string first = req.matches[2];
    	string last = req.matches[3];
    	string phone = req.matches[4];
    	string type = req.matches[5];
    	string email = req.matches[6];
    	string age = req.matches[7];
    	ctdb.editEntry(ID,first,last,phone,type,email,age);

    	res.set_content("{\"status\":\"success\"}", "text/json");
    	res.status = 200;
  	}); 

  	svr.Get(R"(/contact/delete/(.*))", [&](const httplib::Request& req, httplib::Response& res) {
    	res.set_header("Access-Control-Allow-Origin","*");

    	string ID = req.matches[1];
		ctdb.deleteEntry(ID);
    	res.set_content("{\"status\":\"success\"}", "text/json");
    	res.status = 200;
  	});  
  	 
  	cout << "Server listening on port " << port << endl;
  	svr.listen("0.0.0.0", port);

}

File: contactApp.js
// JavaScript for Contact Application Demo Program
// Jim Skon, Kenyon College, 2022
var contactList = [];
const baseUrl = 'http://3.87.226.20:5004';

/* Set up events */
$(document).ready(function() {
	// Add a click event for the filter button
	document.querySelector("#filter-btn").addEventListener("click", (e) => {
		var searchStr = 	document.querySelector("#search").value;
    	findMatches(searchStr);
	});

    $(".dropdown-menu li a").click(function() {
        var selection = $(this).text();
        $(this).parents(".btn-group").find('.btn').html(selection + ' <span class="caret"></span>');
    });

	findMatches(" ");

});

/* Search and display contact functions */

// Build output table from comma delimited list
function formatMatches(json) {

    var result = '<table class="table table-success table-striped""><tr><th>First</th><th>Last</th><th>Phone</th><th>Type</th><th>Action</th><tr>';
    json.forEach(function(entry, i) {
        result += "<tr><td class='first'>" + entry['first'] + "</td><td class='last'>" + entry['last'];
        result += "</td><td class='phone'>" + entry['phone'] + "</td><td class='type'>" + entry['type'] + "</td>";
        result += "<td><button type='button' class='btn btn-primary btn-sm edit' data-bs-toggle='modal' data-bs-target='#editContact' ";
        result += "onclick=\"editContact(" + i + ")\">Edit</button> ";
        result += "<button type='button' class='btn btn-primary btn-sm ' onclick=\"deleteContact("+ entry['ID'] +")\">Delete</button></td></tr>";
    });
    result += "</table>";

    return result;
}

function displayMatches(results) {

    contactList = results["results"];
    console.log("Results:"+JSON.stringify(contactList));
    document.getElementById("searchresults").innerHTML = formatMatches(contactList);
    
}

function findMatches(search) {
	// only include / after find if search not empty
	search = search.trim();
    if (search != "") search = "/" + search;

	console.log("Search:" + search);
    fetch(baseUrl + '/contact/find' + search, {
            method: 'get'
        })
        .then(response => response.json())
        .then(json => displayMatches(json))
        .catch(error => {
            {
                alert("Find Error: Something went wrong:" + error);
            }
        })
}

/* Add contact functions */
function processAdd(results) {
    console.log("Add:", results["status"]);
    document.getElementById("addfirst").value = "";
    document.getElementById("addlast").value = "";
    document.getElementById("addphone").value = "";
    document.getElementById("addemail").value = "";
    document.getElementById("addage").value = "";
    findMatches(" ");

}

function addContact() {
    console.log("Attempting to add an entry");
    console.log("Firstname:" + $('#addfirst').val());
    $('#searchresults').empty();
    fetch(baseUrl + '/contact/add/' + $('#addfirst').val() + "/" + $('#addlast').val() + "/" + $('#addphone').val() + "/" + $('#addtype').text()+ "/" + $('#addemail').val()+ "/" + $('#addage').val(), {
            method: 'get'
        })
        .then(response => response.json())
        .then(json => processAdd(json))
        .catch(error => {
            {
                alert("Add Error: Something went wrong:" + error);
            }
        })
}


function editContact(row) {
    console.log("start edit data: "+row+JSON.stringify(contactList[row]));

    console.log("First name of record: " + contactList[row]["first"] + " " + contactList[row]["last"]);
    editid = contactList[row]["ID"];

	document.getElementById("editfirst").value = contactList[row]["first"];
	document.getElementById("editlast").value = contactList[row]["last"];
	document.getElementById("editphone").value = contactList[row]["phone"];
	document.getElementById("edittype").innerText = contactList[row]["type"];
	document.getElementById("editemail").innerText = contactList[row]["email"];
	document.getElementById("editage").innerText = contactList[row]["age"];
	
	//Save ID in modal
	var modal = document.querySelector("#editContact");
	modal.setAttribute("editid",editid);

}


function updateContact() {

	// Get ID in the modal
	var modal = document.querySelector("#editContact");
	id = modal.getAttribute("editid");
	
    console.log("Attempting to edit an entry:"+id); 

    fetch(baseUrl + '/contact/update/' + id + '/' + document.getElementById("editfirst").value 
    		+ '/' + document.getElementById("editlast").value + '/' + document.getElementById("editphone").value + '/' + document.getElementById("edittype").innerText + '/' + document.getElementById("editemail").value+ '/' + document.getElementById("editage").value, {
                method: 'get'
            })
        .then(alert("Record for " + document.getElementById("editfirst").value + ' ' + document.getElementById("editlast").value + " updated"))
        .catch(error => {
            {
                alert("Edit Error: something went wrong:" + error);
            }
        });
        
    findMatches(" ");

}


function deleteContact(id) {

    console.log("Attempting to delete an entry:" + id);
    fetch(baseUrl + '/contact/delete/' + id, {
            method: 'get'
        })
        .then(alert("Deleted Record: " + id))
        .catch(error => {
            {
                alert("Delete Error: Something went wrong:" + error);
            }
        });
     findMatches(" ");

}


